### ISSUE TITLE
getConfMatrix() on ResampledPrediction: is it majority voting?

### ISSUE THREAD

**Issue by enricoferrero** [mlr-org/mlr#981]
Quick question: When usng `getConfMatrix()` on a `ResampledPrediction` object, how are the numbers calculated? The different models obtained during resampling may well predict some observations in different classes, so how is the aggregation performed? Is it by majority voting?

Thanks!

**Comment by berndbischl** [comment]
if you look at the code this is used to build the table

```
  resp = pred$data$response
  truth = pred$data$truth
```

thats just ALL predicted test data, for all test sets. there is no real aggregation. its simply for every obs you predict you check against it real class. and add a number to the table.

does this help?

i can add a sentence to the docs to clear this up?

**Comment by enricoferrero** [comment]
Yes, that makes sense, for some reason I was assuming the resampling was done with replacement, which is not the case (and therefore there could have been observations predicted with different outcomes, which does not happen in reality).

Related to this, can you please check that the -SUM- columns are correct in this example:

```{r}
library(mlr)
data(iris)

## Define the task
task = makeClassifTask(id = "tutorial", data = iris, target = "Species")

## Define the learner
lrn = makeLearner("classif.lda")

## Define the resampling strategy
rdesc = makeResampleDesc(method = "CV", stratify = TRUE)

## Do the resampling
r = resample(learner = lrn, task = task, resampling = rdesc, show.info = FALSE)

getConfMatrix(r$pred)
#            predicted
#true         setosa versicolor virginica -SUM-
#  setosa         50          0         0     0
#  versicolor      0         48         2     2
#  virginica       0          2        48     2
#  -SUM-           0          2         2     4
```

What are the -SUM- columns counting?

**Comment by berndbischl** [comment]
> Yes, that makes sense, for some reason I was assuming the resampling was done with replacement, which is not the case (and therefore there could have been observations predicted with different outcomes, which does not happen in reality).

well, resampling COULD in principle be with replacement. that depends on your ResampleDesc you chose! eg bootstrap would be with replacement. but that replacement is only done for the training set.
the test sets INDIVIDUALLY dont contain duplicates, for BS. but of course somebody could create a resampling strategy where such duplicated occur in a test set.

also note: in BS and subsampling the test sets are not necessarily disjoint! so table is likely "counting some objects multiple times" in such cases. i do admit the table makes most sense if you have used a CV where then yu have exactly one (y, yhat) pair for each observation in your data set.

 

> What are the -SUM- columns counting?

well here i can quote docs:

#' The marginal elements count the number of classification
#' errors for the respective row or column, i.e., the number of errors
#' when you condition on the corresponding true (rows) or predicted
#' (columns) class. The last element in the margin diagonal
#' displays the total amount of errors.

**Comment by enricoferrero** [comment]
Excellent thanks!

**Comment by berndbischl** [comment]
i added a sentence to the docs about your questions wrt resampling
